README.md

```diff
@@ -1 +1,115 @@
-# runner
\ No newline at end of file
+# runner design doc
+
+## Background
+This repo is intended to hold the client, server (api), and underlying (library) golang code for a gRPC-based runner of linux commands. 
+
+(phew ... that's a mouthful)
+
+Not only that, but it should be secured by TLS, honor cgroups and, optionally, linux namespaces, and stream data back to the client from the running process. 
+Impossible you say?! Maybe. Think of it as making docker runtime without the image management.
+
+This document will outline my approach to make this a reality. 
+
+### Broad strokes
+
+I hope to implement this in a series of PR's that follow this schedule:
+
+0. this doc
+1. create the library for the process runner and basic authz.
+2. basic TLS-secured gRPC client-server chat. The API will wrapper the process runner library
+3. support for hard coded gRPC streaming data. Something simple like a clock tick
+4. Spawn some basic jobs, capture stderr/stdout and pump them into the streams
+5. Work on job control. Cancelling a process, ensuring the channels/goroutines/resources clean up correctly, etc. 
+6. Add cgroups and resource isolation 
+
+(this may change)
+
+Stretch goals
+
+7. some key tests
+8. docker packaging
+9. a nice client
+10. magic
+
+## API
+
+Exposed gRPC "Runner" service methods:
+```
+Auth(username) (Session, error)
+Run(Session, cmd, args...) (Process, error)
+Stdout(Session, Process) (Stream, error)
+Status(Session, Process) (Status, error)
+Abort(Session, Process) (Status, error)
+```
+
+where: 
+ * the `Auth` method returns a `Session` method to authenticated callers.
+ * the `Run` method that can launch new processes for authenticated callers.
+   * the "role" of the authenticated user will determine the cgroup the process runs in
+   * `Run` takes `cmd` and `args` args for the intended linux command and related cmdline parameters and returns a `Process` message
+ * `Process` is the process ID for the process control calls: `Stdout`, `Status`, and `Abort`
+ * `Status` will return an enum of `COMPLETED|RUNNING|ERROR` and, ideally, the exit code
+
+## AuthN / AuthZ
+
+A hardcoded set of usernames (no passwords) will be mapped to a set of V2 cgroups 
+
+| cgroup name | CPU | Memory | 
+| ------------| --- | ------ |
+| **good**    | x   | x      |
+| **better**  | x   | x      |
+| **best**    | x   | x      |
+
+| username | assigned cgroup | 
+| -------- | --------------- |
+| **good** | good            |
+| **better** | better        |
+| **best** | best            |
+
+The caller can log in with the `Auth()` method using one of the whitelisted usernames. Other usernames will result in an error. 
```

Comment:
Your interpretation is correct. The purpose of the separate `Auth` call was to determine the "role" the user operates as. This will determine the good/better/best cgroup they run under. I'm not sure I can keep that same mechanism with a single mTLS cert?

Reply:
Who said you need to use a single mTLS cert? 😀

Can we have different certs for different clients and use the information in the certificate to determine which cgroup that client gets?

Reply:
good point. I'll play with that. Thanks